pkg/pillar/cmd/vcomlink/protoutil.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2018-2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package vcomlink
+
+import (
+	"fmt"
+
+	"github.com/lf-edge/eve/pkg/pillar/vcom"
+	pw "google.golang.org/protobuf/encoding/protowire"
+	pr "google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// according to https://protobuf.dev/programming-guides/encoding/ and
+// go protobuf implementation.
+var wireTypes = map[pr.Kind]pw.Type{
+	pr.BoolKind:     pw.VarintType,
+	pr.EnumKind:     pw.VarintType,
+	pr.Int32Kind:    pw.VarintType,
+	pr.Sint32Kind:   pw.VarintType,
+	pr.Uint32Kind:   pw.VarintType,
+	pr.Int64Kind:    pw.VarintType,
+	pr.Sint64Kind:   pw.VarintType,
+	pr.Uint64Kind:   pw.VarintType,
+	pr.Sfixed32Kind: pw.Fixed32Type,
+	pr.Fixed32Kind:  pw.Fixed32Type,
+	pr.FloatKind:    pw.Fixed32Type,
+	pr.Sfixed64Kind: pw.Fixed64Type,
+	pr.Fixed64Kind:  pw.Fixed64Type,
+	pr.DoubleKind:   pw.Fixed64Type,
+	pr.StringKind:   pw.BytesType,
+	pr.BytesKind:    pw.BytesType,
+	pr.MessageKind:  pw.BytesType,
+	pr.GroupKind:    pw.StartGroupType,
+}
+
+// extractBaseFromPacket basically extracts the channel from the packet and
+// populates the Base message. The reason we are not using proto.Unmarshal is
```

Comment:
Oh.... Now it looks super complicated... I even regret I suggested protobuf...

Actually, why do you need to read the message partially? You need to get the type first (channel in your case)?

If so, wouldn't it be simpler to use just a wrapper-approach?

```proto
syntax = "proto3";

package vcom;
option go_package = ".;vcom";

enum Channels {
    CHANNEL_UNSPECIFIED = 0;
    ERROR = 1;
    TPM = 2;
}

enum TpmRequests {
    REQUEST_UNSPECIFIED = 0;
    GET_EK = 1;
}

// Base is the base packet for all other packets
message Base {
    uint32 channel = 1;
    bytes payload = 2; // For the rest of the message
}

message Error {
    string error = 1;
}

message TpmRequest {
    uint32 request = 1;
    // add extra fields as needed
}

// TpmResponseEk is the response packet for TPM_GET_EK
message TpmResponseEk {
    string ek = 1;
}
```

And then you can marshal/unmarshal like this...
```go
package vcomlink

import (
    "fmt"

    "github.com/golang/protobuf/proto"
    "github.com/lf-edge/eve/pkg/pillar/vcom"
)

// EncodeMessage wraps the actual message in a Base message.
func EncodeMessage(msg proto.Message, channel uint32) ([]byte, error) {
    payload, err := proto.Marshal(msg)
    if err != nil {
        return nil, fmt.Errorf("failed to marshal message: %w", err)
    }

    base := &vcom.Base{
        Channel: channel,
        Payload: payload,
    }

    return proto.Marshal(base)
}

// DecodeMessage unwraps the Base message and decodes the actual message.
func DecodeMessage(data []byte) (proto.Message, uint32, error) {
    var base vcom.Base
    if err := proto.Unmarshal(data, &base); err != nil {
        return nil, 0, fmt.Errorf("failed to unmarshal base: %w", err)
    }

    var msg proto.Message
    switch base.Channel {
    case uint32(vcom.Channels_ERROR):
        msg = &vcom.Error{}
    case uint32(vcom.Channels_TPM):
        msg = &vcom.TpmRequest{}
    // Add cases for other message types as needed
    default:
        return nil, 0, fmt.Errorf("unknown channel: %d", base.Channel)
    }

    if err := proto.Unmarshal(base.Payload, msg); err != nil {
        return nil, 0, fmt.Errorf("failed to unmarshal payload: %w", err)
    }

    return msg, base.Channel, nil
}
```

It will avoid the complexity... You can just read until the end of data and decode it...
I think you get the idea.

Reply:
To be honest, I liked the json one better, simple, partial decoding, human readable... 

I get your idea, but I think it makes it even uglier than it already is 😄 my goal was that if a client want to send a TpmRequets, it instantiates a TpmRequest, fill it, marshal it and send it, but now they have to instantiate two different messages and wrap one into another. 

I can try flatbuffers, I've heard that one allows partial decoding (it is just a flat buffer so makes sense).

Reply:
@OhmSpectator I tried flatbuffers, oh the humanity! it is ugly and full of boilerplate. For example this is the code for just serializing a simple error message based on our schema : 

```
func encodeError(msg string) []byte {
	builder := fbs.NewBuilder(0)

	vcom.BaseStart(builder)
	vcom.BaseAddChannel(builder, vcom.ChannelsERROR)
	base := vcom.BaseEnd(builder)

	vcom.ErrorStart(builder)
	vcom.ErrorAddBase(builder, base)
	vcom.ErrorAddError(builder, builder.CreateString(msg))
	errorMsg := vcom.ErrorEnd(builder)
	builder.Finish(errorMsg)

	return builder.FinishedBytes()
}
```

and the when I want to deserialize say TpmRequest, this is what I do, mind it is hierarchical so I suppose we need to know the offset (how?!), here I just set to 0: 

```
obj := vcom.GetRootAsTpmRequest(buf, 0)
```

and I don't even know how to check for errors! it seems like it panics if the message is for example to short, see the internals: 

```
func GetRootAsTpmRequest(buf []byte, offset flatbuffers.UOffsetT) *TpmRequest {
	n := flatbuffers.GetUOffsetT(buf[offset:])
	x := &TpmRequest{}
	x.Init(buf, n+offset)
	return x
}

func GetUOffsetT(buf []byte) UOffsetT {
	return UOffsetT(GetUint32(buf))
}

func (rcv *TpmRequest) Init(buf []byte, i flatbuffers.UOffsetT) {
	rcv._tab.Bytes = buf
	rcv._tab.Pos = i
}

func (rcv *TpmRequest) Request() TpmRequests {
	o := flatbuffers.UOffsetT(rcv._tab.Offset(6))
	if o != 0 {
		return TpmRequests(rcv._tab.GetUint32(o + rcv._tab.Pos))
	}
	return 0
}
```

I'm sure there is error checking, but is it? щ(ಥДಥщ)